docs/RFCS/20191028_easier_hash_sharded_indexes.md

```diff
@@ -0,0 +1,220 @@
+time period. We might start with the following schema.
+
+```sql
+CREATE TABLE events (
+    device_id
+        UUID,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (device_id, ts, event_id),
+    INDEX (ts)
+);
+```
+
+This schema would have a hot spot on that INDEX (ts) which would be rather unfortunate. We
+can alleviate this hot spot by sharding this time ordered index.
+
+```sql
+CREATE TABLE events (
+    device_id
+        UUID,
+    shard
+        INT8
+        AS (fnv32(device_id) % 8) STORED
+        CHECK (shard IN (0, 1, 2, 3, 4, 5, 6, 7)),
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (device_id, ts, event_id),
+    INDEX (shard, ts)
+);
+```
+
+This isn't too big of a lift here because the device ID is easy to hash in sql. Imagine
+instead we had a primary key based on some other features:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts)
+);
+```
+
+In order to shard this we'll need something like:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    shard
+        INT8
+        AS (
+            fnv32(
+                concat(hex(product_id)),
+                owner::STRING,
+                serial_number
+            )
+            % 8
+        ) STORED
+        CHECK (shard IN (0, 1, 2, 3, 4, 5, 6, 7)),
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (
+        product_id,
+        owner,
+        serial_number,
+        ts,
+        event_id
+    ),
+    INDEX (shard, ts)
+);
+```
+
+We can see that this is starting to get heavy. The proposal is that we shoulder the burden
+of hashing and installing a check constraint behind a new syntax:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts) USING HASH (product_id, owner, serial_number) WITH (BUCKET_COUNT=8)
+);
+```
+
+# Benchmarks 
+
+ We benchmarked the sequential kv workload against two schemas:
+
+ (1) Using hash sharded primary index on `k`.
+```sql
+CREATE TABLE kv (
+    k INT8 NOT NULL,
+    v bytes NOT NULL,
+    shard 
+        INT8 AS (k % 8) stored 
+        CHECK (shard IN (0, 1, 2, 3, 4, 5, 6, 7)), 
+    PRIMARY KEY (shard, k)
+);
+```
+
+(2) Using primary index on `k`.
+```sql
+CREATE TABLE kv (
+    k INT8 NOT NULL,
+    v BYTES NOT NULL,
+    PRIMARY KEY (k)
+);
+```
+
+ The following benchmarks were all run on a 7 node GCE cluster with `n1-standard-32`
```

Comment:
These are big machines, you should consider more shards. Play around with something like one shard per core or maybe a little bit less? Perhaps 64? Alternatively try smaller machines.

Reply:
okay, smaller machines seem like a better idea, I'll regenerate these numbers.